src/ProtocolGateway.Core/Mqtt/MqttAdapter.cs

```diff
@@ -375,61 +383,121 @@ async void ProcessPendingSubscriptionChanges(IChannelHandlerContext context)
 
         #region PUBLISH Client -> Server handling
 
-        async Task PublishToServerAsync(IChannelHandlerContext context, IMessagingServiceClient sendingClient, PublishPacket packet, string messageType)
+        async Task PublishBatchToServerAsync(IChannelHandlerContext context, IBatchAwareMessagingServiceClient sendingClient, IReadOnlyCollection<PublishPacket> packets, string messageType)
         {
             if (!this.ConnectedToService)
             {
-                packet.Release();
+                foreach (var packet in packets)
+                {
+                    packet.Release();
+                }
                 return;
             }
 
             PreciseTimeSpan startedTimestamp = PreciseTimeSpan.FromStart;
 
             this.ResumeReadingIfNecessary(context);
 
-            IMessage message = null;
+            List<IMessage> messages = new List<IMessage>(packets.Count);
             try
             {
-                message = sendingClient.CreateMessage(packet.TopicName, packet.Payload);
-                Util.CompleteMessageFromPacket(message, packet, this.settings);
-
-                if (messageType != null)
+                foreach (var packet in packets)
                 {
-                    message.Properties[this.settings.ServicePropertyPrefix + MessagePropertyNames.MessageType] = messageType;
+                    IMessage message = this.ConvertPacketToMessage(sendingClient, packet, messageType);
+                    messages.Add(message);
                 }
 
-                await sendingClient.SendAsync(message);
+                await sendingClient.SendBatchAsync(messages);
 
-                PerformanceCounters.MessagesSentPerSecond.Increment();
+                messages.Clear();
+
+                PerformanceCounters.MessagesSentPerSecond.IncrementBy(messages.Count);
```

Comment:
nit:If you clear the items above would it not be 0?

Reply:
May be you wanted capacity ?

Reply:
no, cleaning after is the right thing to do.